File: android/app/src/main/java/com/voicerewriter/textproc/ListFormatter.kt
package com.voicerewriter.textproc

/**
 * Turns explicit spoken structure into real newlines:
 *  - "new line" -> "\n", "new paragraph" -> "\n\n".
 *  - explicit numbered enumerations ("for 1. apples 2. bananas 3. oranges",
 *    "first … second … third …", or "one … two … three …") -> a newline-separated list.
 *
 * Conservative on purpose: a list only forms from a run of enumerators that starts
 * at 1/"first"/"one" and increases consecutively (≥2 for digit markers, ≥3 for the
 * ambiguous ordinal/cardinal words). Spoken-word enumerators are matched at any word
 * boundary ("three things first the api second the docs …"), not only after
 * punctuation, since the STT transcript often doesn't punctuate the lead-in. Natural
 * enumerations without markers ("apples, bananas and oranges") are left for the LLM.
 *
 * Runs after NumberNormalizer so digits are settled and no earlier stage flattens
 * the newlines it introduces.
 */
object ListFormatter {

    private val ordinals = linkedMapOf(
        "first" to 1, "second" to 2, "third" to 3, "fourth" to 4, "fifth" to 5,
        "sixth" to 6, "seventh" to 7, "eighth" to 8, "ninth" to 9, "tenth" to 10,
    )

    private val cardinals = linkedMapOf(
        "one" to 1, "two" to 2, "three" to 3, "four" to 4, "five" to 5,
        "six" to 6, "seven" to 7, "eight" to 8, "nine" to 9, "ten" to 10,
    )

    fun format(text: String): String {
        if (text.isBlank()) return text
        val withBreaks = convertNewlineMarks(text)
        if (withBreaks.contains('\n')) return withBreaks // explicit breaks win; don't also list-ify
        return formatNumberedList(withBreaks)
    }

    private fun convertNewlineMarks(text: String): String {
        var r = text
        r = Regex("\\bnew\\s+paragraph\\b", RegexOption.IGNORE_CASE).replace(r, "\n\n")
        r = Regex("\\bnew\\s+lines?\\b|\\bnewline\\b", RegexOption.IGNORE_CASE).replace(r, "\n")
        r = r.replace(Regex("[ \\t]*\\n[ \\t]*"), "\n") // tidy spaces around inserted breaks
        return r.trim()
    }

    /** (markerStart, markerEnd, value) for each enumerator found, in text order. */
    private fun formatNumberedList(text: String): String {
        // Digit markers: "1. " / "2) " — separator + following whitespace (so "2.0" / times don't match).
        val digit = Regex("(?<=^|\\s)(\\d+)\\s*[.)]\\s+").findAll(text)
            .map { Triple(it.range.first, it.range.last + 1, it.groupValues[1].toInt()) }.toList()
        buildList(text, digit, minRun = 2)?.let { return it }

        // Spoken-word enumerators: ordinals ("first … second … third") or cardinals
        // ("one … two … three"). Both are ambiguous, so a list only forms from a consecutive
        // run starting at 1 with ≥3 items (and non-empty items — buildList enforces that, so
        // "one, two, three of those" with nothing between markers does NOT become a list).
        for (words in listOf(ordinals, cardinals)) {
            val marks = Regex("(?<=^|\\s)(${words.keys.joinToString("|")})\\b,?\\s+", RegexOption.IGNORE_CASE)
                .findAll(text)
                .mapNotNull { m -> words[m.groupValues[1].lowercase()]?.let { Triple(m.range.first, m.range.last + 1, it) } }
                .toList()
            buildList(text, marks, minRun = 3)?.let { return it }
        }

        return text
    }

    private fun buildList(text: String, markers: List<Triple<Int, Int, Int>>, minRun: Int): String? {
        if (markers.size < minRun) return null
        val start = markers.indexOfFirst { it.third == 1 }
        if (start < 0) return null
        var end = start
        var expected = 2
        var j = start + 1
        while (j < markers.size && markers[j].third == expected) { end = j; expected++; j++ }
        if (end - start + 1 < minRun) return null

        val run = markers.subList(start, end + 1)
        val sb = StringBuilder()
        val leadIn = text.substring(0, run.first().first).trim()
        if (leadIn.isNotEmpty()) sb.append(leadIn).append("\n")
        for ((k, m) in run.withIndex()) {
            val itemEnd = if (k + 1 < run.size) run[k + 1].first else text.length
            val item = text.substring(m.second, itemEnd).trim().trimEnd(',', ';')
            if (item.isEmpty()) return null
            sb.append("${m.third}. ").append(item)
            if (k < run.size - 1) sb.append("\n")
        }
        return sb.toString()
    }
}

File: macos/OpenWisprCore/Sources/OpenWisprCore/ListFormatter.swift
import Foundation

/// Turns explicit spoken structure into real newlines and numbered lists.
/// Port of the Kotlin `ListFormatter`. (Shadows Foundation's `ListFormatter`
/// within this module; reference as `OpenWisprCore.ListFormatter` if disambiguation
/// is needed.)
public enum ListFormatter {

    /// Ordinal word -> value, in insertion order (mirrors Kotlin linkedMapOf).
    private static let ordinalsOrder: [(String, Int)] = [
        ("first", 1), ("second", 2), ("third", 3), ("fourth", 4), ("fifth", 5),
        ("sixth", 6), ("seventh", 7), ("eighth", 8), ("ninth", 9), ("tenth", 10),
    ]
    private static let ordinals: [String: Int] = Dictionary(uniqueKeysWithValues: ordinalsOrder)

    public static func format(_ text: String) -> String {
        if text.ktIsBlank { return text }
        let withBreaks = convertNewlineMarks(text)
        if withBreaks.contains("\n") { return withBreaks } // explicit breaks win
        return formatNumberedList(withBreaks)
    }

    private static func convertNewlineMarks(_ text: String) -> String {
        var r = text
        r = KRegex("\\bnew\\s+paragraph\\b", ignoreCase: true).replace(r, "\n\n")
        r = KRegex("\\bnew\\s+lines?\\b|\\bnewline\\b", ignoreCase: true).replace(r, "\n")
        r = KRegex("[ \\t]*\\n[ \\t]*").replace(r, "\n") // tidy spaces around inserted breaks
        return r.ktTrim()
    }

    /// (markerStart, markerEnd, value) for each enumerator found, in text order.
    private static func formatNumberedList(_ text: String) -> String {
        // Digit markers: "1. " / "2) ".
        let digit = KRegex("(?<=^|\\s)(\\d+)\\s*[.)]\\s+").findAll(text)
            .map { (start: $0.start, end: $0.end, value: Int($0.groupValues[1])!) }
        if let r = buildList(text, digit, minRun: 2) { return r }

        // Ordinal words at a clause start. Ambiguous → require ≥3.
        let ordKeys = ordinalsOrder.map { $0.0 }.joined(separator: "|")
        let ords = KRegex("(?<=^|[.,;:]\\s)(\(ordKeys))\\b,?\\s+", ignoreCase: true)
            .findAll(text)
            .compactMap { m -> (start: Int, end: Int, value: Int)? in
                guard let v = ordinals[m.groupValues[1].lowercased()] else { return nil }
                return (start: m.start, end: m.end, value: v)
            }
        if let r = buildList(text, ords, minRun: 3) { return r }

        return text
    }

    private static func buildList(_ text: String, _ markers: [(start: Int, end: Int, value: Int)], minRun: Int) -> String? {
        if markers.count < minRun { return nil }
        guard let start = markers.firstIndex(where: { $0.value == 1 }) else { return nil }
        var end = start
        var expected = 2
        var j = start + 1
        while j < markers.count && markers[j].value == expected { end = j; expected += 1; j += 1 }
        if end - start + 1 < minRun { return nil }

        let run = Array(markers[start...end])
        let ns = text as NSString
        var sb = ""
        let leadIn = ns.substring(to: run.first!.start).ktTrim()
        if !leadIn.isEmpty { sb += leadIn + "\n" }
        for (k, m) in run.enumerated() {
            let itemEnd = (k + 1 < run.count) ? run[k + 1].start : ns.length
            let item = ns.substring(with: NSRange(location: m.end, length: itemEnd - m.end))
                .ktTrim().ktTrimEnd(",", ";")
            if item.isEmpty { return nil }
            sb += "\(m.value). " + item
            if k < run.count - 1 { sb += "\n" }
        }
        return sb
    }
}
